src/main/java/com/uid2/admin/secret/SaltRotation.java

```diff
@@ -94,86 +125,142 @@ private String calculatePreviousSalt(SaltEntry salt, boolean shouldRotate, long
+            indexesToRotate.addAll(saltsToRotate);
+            minLastUpdated = maxLastUpdated;
         }
         return indexesToRotate;
     }
 
-    private SaltEntry[] getRotatableSalts(SaltSnapshot lastSnapshot, long nextEffective) {
-        SaltEntry[] salts = lastSnapshot.getAllRotatingSalts();
-        if (isRefreshFromEnabled) {
-            return Arrays.stream(salts).filter(s -> s.refreshFrom() == nextEffective).toArray(SaltEntry[]::new);
+    private List<Integer> selectIndexesToRotate(
+            SaltEntry[] salts,
+            long minLastUpdated,
+            long maxLastUpdated,
+            int maxIndexes,
+            List<Integer> rotatableSaltIndexes
+    ) {
+        var candidateIndexes = indexesForRotation(salts, minLastUpdated, maxLastUpdated, rotatableSaltIndexes);
+
+        if (candidateIndexes.size() <= maxIndexes) {
+            return candidateIndexes;
         }
-        return salts;
+        Collections.shuffle(candidateIndexes);
+        return candidateIndexes.subList(0, Math.min(maxIndexes, candidateIndexes.size()));
     }
 
+    private List<Integer> indexesForRotation(
+            SaltEntry[] salts,
+            long minLastUpdated,
+            long maxLastUpdated,
+            List<Integer> rotatableSaltIndexes
+    ) {
+        var candidateIndexes = new ArrayList<Integer>();
+        for (int i = 0; i < salts.length; i++) {
+            var salt = salts[i];
+            var lastUpdated = salt.lastUpdated();
+            var isInTimeWindow = minLastUpdated <= lastUpdated && lastUpdated < maxLastUpdated;
+            var isRotatable = rotatableSaltIndexes.contains(i);
 
-    private void addIndexesToRotate(List<Integer> entryIndexes,
-                                    SaltEntry[] entries,
-                                    long minLastUpdated,
-                                    long maxLastUpdated,
-                                    int maxIndexes) {
-        final List<Integer> candidateIndexes = IntStream.range(0, entries.length)
-                .filter(i -> isBetween(entries[i].lastUpdated(), minLastUpdated, maxLastUpdated))
-                .boxed()
-                .collect(toList());
-        if (candidateIndexes.size() <= maxIndexes) {
-            entryIndexes.addAll(candidateIndexes);
-            return;
+            if (isInTimeWindow && isRotatable) {
+                candidateIndexes.add(i);
+            }
         }
-        Collections.shuffle(candidateIndexes);
-        candidateIndexes.stream().limit(maxIndexes).forEachOrdered(entryIndexes::add);
+        return candidateIndexes;
     }
 
-    private static boolean isBetween(long t, long minInclusive, long maxExclusive) {
-        return minInclusive <= t && t < maxExclusive;
+    private boolean isRotatable(long nextEffective, SaltEntry salt) {
+        if (this.isRefreshFromEnabled) {
+            if (salt.refreshFrom() == null) { // TODO: remove once refreshFrom is no longer optional
```

Comment:
Is it still optional? I thought we've populated this everywhere now?

Reply:
Yeah, in practice we have that populated in every environment, there are still code paths for null case though. I'll remove it from here, we can clean up the others later.